Use with blocks (or try/finally) to ensure filehandles are closed

terminalmage opened the pull request:
This fixes a few sources (or potential sources) of filehandle leaks.

It also fixes several PY3 test issues.

terminalmage committed:
Use with blocks (or try/finally) to ensure filehandles are closed

This fixes a few sources (or potential sources) of filehandle leaks.

rallytime commented on salt/proxy/dummy.py:
```diff
@@ -35,16 +35,14 @@ def __virtual__():
-    pck = open(FILENAME, 'wb')  # pylint: disable=W8470
-    pickle.dump(details, pck)
-    pck.close()
+    with open(FILENAME, 'wb') as pck:
```
These all need to use the `salt.utils.files.fopen` function instead of Python's `open`.

rallytime commented on salt/proxy/dummy.py:
```diff
@@ -35,16 +38,14 @@ def __virtual__():
-    pck = open(FILENAME, 'wb')  # pylint: disable=W8470
-    pickle.dump(details, pck)
-    pck.close()
+    with salt.utils.fopen(FILENAME, 'wb') as pck:
```
This line needs to use the new path `salt.utils.files.fopen` instead of `salt.utils.fopen`.

rallytime commented on salt/proxy/dummy.py:
```diff
@@ -35,16 +38,14 @@ def __virtual__():
-        pck = open(FILENAME, 'r')  # pylint: disable=W8470
-        DETAILS = pickle.load(pck)
-        pck.close()
+        with salt.utils.fopen(FILENAME, 'r') as pck:
```
Same comment here.

rallytime reviewed (changes requested):
@terminalmage There are two comments that need to be resolved. Then we can get this in.

terminalmage committed:
Fix stacklevel in salt/utils/__init__.py

terminalmage committed:
Wait until after restoring original resource limit before removing temp files

Invoking shutil.rmtree() first was causing a traceback since the max
open files setting was still set very low for the test. This moves the
shutil.rmtree() until after we have restored the original resource
limit.

terminalmage committed:
Python 3 compatibility fixes for kubernetes

dict.keys() is not a list in Python 3, and base64.b64encode requires
a bytestring.

terminalmage commented:
I've also fixed the kubernetes tests that were failing on Python 3.

terminalmage committed:
Update reference to old salt.utils.which function

terminalmage commented:
Made one more commit fixing a reference to one of the salt.utils funcs which was moved in develop. This should be GTM pending the results of the tests currently running.

rallytime commented:
You rock @terminalmage! Thanks for fixing these.
